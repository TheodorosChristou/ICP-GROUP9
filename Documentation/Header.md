# Header Component

This component represents the header/navigation bar of the application.

## Example Usage

```jsx
import Header from './Header';

const YourComponent = () => {
  return (
    <div>
      <Header />
      {/* Your content */}
    </div>
  );
};

export default YourComponent;
```

## Component Structure

```tsx
<Header>
  <nav data-testid="navbar" className="...">
    {/* Logo and title */}
    <div className="...">
      <div className="...">
        <Link href="/">
          <Image alt="Maritime Logo" src="/img/Anchor.png" width={60} height={60} className="..." />
        </Link>
      </div>
      <div className="...">
        <Link href="/">
          <h1>Maritime</h1>
        </Link>
      </div>
      <div className="...">
        <h1>Emergency</h1>
      </div>
    </div>
    {/* Menu toggle button */}
    <div className="...">
      <button onClick={toggleMenu} className="...">
        {menuOpen ? <XIcon className="..." /> : <MenuIcon className="..." />}
      </button>
    </div>
    {/* Menu items */}
    <div className={`... ${menuOpen ? '...' : 'hidden'}`}>
      <Link href="/" passHref>
        <div className="..." onClick={handleMenuClick}>Home</div>
      </Link>
      <Link href="/map" passHref>
        <div className="..." onClick={handleMenuClick}>Map</div>
      </Link>
      {session && (role === 'admin' || process.env.NEXT_PUBLIC_TESTING) && (
        <Link href="/archive" passHref>
          <div className="..." onClick={handleMenuClick}>Archive</div>
        </Link>
      )}
      {!session && (
        <div className="...">
          <div onClick={() => signIn()} className="..."><LoginIcon className="..." /> Log In</div>
        </div>
      )}
      {session && (
        <div className="...">
          <div className="...">Welcome {session.user.name}</div>
          <div onClick={() => signOut()} className="..."><LogoutIcon className="..." /> Log Out</div>
        </div>
      )}
    </div>
  </nav>
</Header>

```

## Props

This component does not accept any props.

## Implementation Details

- Uses useSession hook from next-auth/react to get user session data.

- Renders different content based on user authentication status and role.

- Provides a toggle button to show/hide the menu on smaller screens.

- Uses Link from Next.js for client-side navigation.