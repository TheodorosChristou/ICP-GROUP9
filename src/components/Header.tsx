import {LoginIcon, LogoutIcon} from "@heroicons/react/outline";
import Image from "next/image";
import Link from "next/link";
import {signIn, signOut, useSession} from "next-auth/react"


export default function Header() {

  const{data: session} = useSession();

  return (
    <div className="fixed top-0 left-0 right-0 z-40 bg-white">
      <nav className="p-2 items-center w-full flex justify-between">
        <div className="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between">
          <div className="flex align-center">
            <div className="flex space-x-2">
              <Link href="/">
                <Image
                  alt="image"
                  src="/img/iq.png"
                  width={60}
                  height={60}
                  className="cursor-pointer"
                />
              </Link>
            </div>
            <div className="flex space-x-2 ml-1.5 mt-1.5 text-black	 mb-1 font-bold italic text-3xl py-1 px-1 xs:px-3 sm:px-1"><a href="/"><h1>EuroTech</h1></a></div>
            <div className="flex text-black space-x-2 cursor-default mt-3 mb-1 font-semibold italic text-xl py-1 sm:px-1"><h1>Tasker</h1></div>
          </div>
          <div className="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end">
            <div className="mx-2.5 gap-1 hidden lg:flex">
              <div className="flex"><a className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full" href="/">Home</a></div>
              <div className="flex"><a className="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full" href="/archive">Archive</a></div>
            </div>
            {!session &&(<div className="flex items-center mb-2 mx-2.5">
              <div><h1 className="text-5xl cursor-default mb-1">|</h1></div>
              <div className="flex cursor-pointer" onClick={()=> signIn()}><a className="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"><LoginIcon className="h-6 w-6 mt-1" /> Log In</a></div>
            </div>)}
            {session &&(<div className="flex items-center mb-2 mx-2.5">
              <div><h1 className="text-5xl cursor-default mb-1">|</h1></div>
              <div className="flex items-center text-xl mt-2 py-1 px-1 xs:px-3 sm:px-3 cursor-default"><h1>Welcome {session.user!.name}</h1></div>
              <div className="flex cursor-pointer"onClick={()=> signOut()}><a className="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"><LogoutIcon className="h-6 w-6 mt-1" /> Log Out</a></div>
            </div>)}
          </div>
          </div>
      </nav>
    </div>
  );
}
