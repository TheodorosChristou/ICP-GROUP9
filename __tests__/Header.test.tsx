import React from 'react'
import {render, screen, fireEvent} from '@testing-library/react'
import Header from '@/components/Header'

jest.mock("next-auth/react", () => {
    const originalModule = jest.requireActual('next-auth/react');
    const mockSession = {
      expires: new Date(Date.now() + 2 * 86400).toISOString(),
      user: { username: "admin" }
    };
    return {
      __esModule: true,
      ...originalModule,
      useSession: jest.fn(() => {
        return {data: mockSession, status: 'authenticated'}  // return type is [] in v3 but changed to {} in v4
      }),
    };
  });

  it('Should have a nav bar', () => {
    render(<Header />);
    const navBar = screen.getByTestId('navbar');
    expect(navBar).toHaveClass('p-2')
    expect(screen.getByText('Home')).toBeInTheDocument();
    expect(screen.getByText('Map')).toBeInTheDocument();
})
it('Should have an image', () => {
    render(<Header />);
    const image = screen.getByAltText('image');
    expect(image).toBeInTheDocument()
})

it('Should Show Log Out when has session',
    async () => {
    const {container} = render(<Header/>);
  
    expect(container).toMatchSnapshot()
    expect(screen.getByText("Log Out")).toBeInTheDocument();
    });

it('Should Show Log In when no session', () => {
    jest.mock("next-auth/react", () => ({
    useSession: () => ({ data: null })
    }));
      
    render(<Header/>);
      
    expect(screen.queryByText("Log In")).not.toBeInTheDocument();
    });
