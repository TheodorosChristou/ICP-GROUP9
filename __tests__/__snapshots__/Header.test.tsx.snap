// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Show Log In when no session 1`] = `
<div>
  <div
    class="fixed top-0 left-0 right-0 z-40 bg-white"
  >
    <nav
      class="p-2 items-center w-full flex justify-between"
      data-testid="navbar"
    >
      <div
        class="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between"
      >
        <div
          class="flex align-center"
        >
          <div
            class="flex space-x-2"
          >
            <a
              href="/"
            >
              <img
                alt="image"
                class="cursor-pointer"
                data-nimg="1"
                decoding="async"
                height="60"
                loading="lazy"
                src="/_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75"
                srcset="/_next/image?url=%2Fimg%2Fanchor.png&w=64&q=75 1x, /_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75 2x"
                style="color: transparent;"
                width="60"
              />
            </a>
          </div>
          <div
            class="flex space-x-2 ml-1.5 mt-1.5 text-black mb-1 font-bold italic text-3xl py-1 px-1 xs:px-3 sm:px-1"
          >
            <a
              href="/"
            >
              <h1>
                Maritime
              </h1>
            </a>
          </div>
          <div
            class="flex text-black space-x-2 cursor-default mt-3 mb-1 font-semibold italic text-xl py-1 sm:px-1"
          >
            <h1>
              Emergency
            </h1>
          </div>
        </div>
        <div
          class="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end"
        >
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/"
              >
                Home
              </a>
            </div>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/map"
              >
                Map
              </a>
            </div>
          </div>
          <div
            class="flex items-center mb-2 mx-2.5"
          >
            <div>
              <h1
                class="text-5xl cursor-default mb-1"
              >
                |
              </h1>
            </div>
            <div
              class="flex items-center text-xl mt-2 py-1 px-1 xs:px-3 sm:px-3 cursor-default"
            >
              <h1>
                Welcome 
              </h1>
            </div>
            <div
              class="flex cursor-pointer"
            >
              <a
                class="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 mt-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                 Log Out
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;

exports[`Should Show Log Out when has session 1`] = `
<div>
  <div
    class="fixed top-0 left-0 right-0 z-40 bg-white"
  >
    <nav
      class="p-2 items-center w-full flex justify-between"
      data-testid="navbar"
    >
      <div
        class="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between"
      >
        <div
          class="flex align-center"
        >
          <div
            class="flex space-x-2"
          >
            <a
              href="/"
            >
              <img
                alt="image"
                class="cursor-pointer"
                data-nimg="1"
                decoding="async"
                height="60"
                loading="lazy"
                src="/_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75"
                srcset="/_next/image?url=%2Fimg%2Fanchor.png&w=64&q=75 1x, /_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75 2x"
                style="color: transparent;"
                width="60"
              />
            </a>
          </div>
          <div
            class="flex space-x-2 ml-1.5 mt-1.5 text-black mb-1 font-bold italic text-3xl py-1 px-1 xs:px-3 sm:px-1"
          >
            <a
              href="/"
            >
              <h1>
                Maritime
              </h1>
            </a>
          </div>
          <div
            class="flex text-black space-x-2 cursor-default mt-3 mb-1 font-semibold italic text-xl py-1 sm:px-1"
          >
            <h1>
              Emergency
            </h1>
          </div>
        </div>
        <div
          class="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end"
        >
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/"
              >
                Home
              </a>
            </div>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/map"
              >
                Map
              </a>
            </div>
          </div>
          <div
            class="flex items-center mb-2 mx-2.5"
          >
            <div>
              <h1
                class="text-5xl cursor-default mb-1"
              >
                |
              </h1>
            </div>
            <div
              class="flex items-center text-xl mt-2 py-1 px-1 xs:px-3 sm:px-3 cursor-default"
            >
              <h1>
                Welcome 
              </h1>
            </div>
            <div
              class="flex cursor-pointer"
            >
              <a
                class="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 mt-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                 Log Out
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;

exports[`Should Show archive when has session 1`] = `
<div>
  <div
    class="fixed top-0 left-0 right-0 z-40 bg-white"
  >
    <nav
      class="p-2 items-center w-full flex justify-between"
      data-testid="navbar"
    >
      <div
        class="text-black h-12 w-full gap-0 sm:gap-3 flex items-center justify-between"
      >
        <div
          class="flex align-center"
        >
          <div
            class="flex space-x-2"
          >
            <a
              href="/"
            >
              <img
                alt="image"
                class="cursor-pointer"
                data-nimg="1"
                decoding="async"
                height="60"
                loading="lazy"
                src="/_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75"
                srcset="/_next/image?url=%2Fimg%2Fanchor.png&w=64&q=75 1x, /_next/image?url=%2Fimg%2Fanchor.png&w=128&q=75 2x"
                style="color: transparent;"
                width="60"
              />
            </a>
          </div>
          <div
            class="flex space-x-2 ml-1.5 mt-1.5 text-black mb-1 font-bold italic text-3xl py-1 px-1 xs:px-3 sm:px-1"
          >
            <a
              href="/"
            >
              <h1>
                Maritime
              </h1>
            </a>
          </div>
          <div
            class="flex text-black space-x-2 cursor-default mt-3 mb-1 font-semibold italic text-xl py-1 sm:px-1"
          >
            <h1>
              Emergency
            </h1>
          </div>
        </div>
        <div
          class="justify-center items-center gap-1 flex 3xl:flex-1 flex-row 3xl:justify-end"
        >
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/"
              >
                Home
              </a>
            </div>
          </div>
          <div
            class="mx-2.5 gap-1 hidden lg:flex"
          >
            <div
              class="flex"
            >
              <a
                class="text-xl hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
                href="/map"
              >
                Map
              </a>
            </div>
          </div>
          <div
            class="flex items-center mb-2 mx-2.5"
          >
            <div>
              <h1
                class="text-5xl cursor-default mb-1"
              >
                |
              </h1>
            </div>
            <div
              class="flex items-center text-xl mt-2 py-1 px-1 xs:px-3 sm:px-3 cursor-default"
            >
              <h1>
                Welcome 
              </h1>
            </div>
            <div
              class="flex cursor-pointer"
            >
              <a
                class="flex items-center text-xl mt-2 hover:bg-gray-300 py-1 px-1 xs:px-3 sm:px-3 rounded-full"
              >
                <svg
                  aria-hidden="true"
                  class="h-6 w-6 mt-1"
                  fill="none"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                  />
                </svg>
                 Log Out
              </a>
            </div>
          </div>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
